src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.CoreCLR.cs

```diff
@@ -171,240 +203,386 @@ private static unsafe object AllocContinuationResultBox(void* ptr)
+                        {
+                            // Can inline
+                            return false;
+                        }
+                    }
+
+                    // Move to thread pool. TODO: Avoid allocation here, but currently we cannot queue `task` directly
+                    // due to assert in ThreadPoolWorkQueue.Enqueue.
+                    ThreadPool.UnsafeQueueUserWorkItemInternal(new ThunkTaskThreadPoolWorkItem<T, TOps>(task), preferLocal: true);
+                    return true;
                 }
-                catch (Exception ex)
+
+                if ((continuation.Flags & CorInfoContinuationFlags.CORINFO_CONTINUATION_CONTINUE_ON_CAPTURED_SYNCHRONIZATION_CONTEXT) != 0)
                 {
-                    continuation = UnwindToPossibleHandler(continuation);
-                    if (continuation.Resume == null)
+                    object continuationContext = continuation.GetContinuationContext();
+                    Debug.Assert(continuationContext is SynchronizationContext { });
+                    SynchronizationContext continuationSyncCtx = (SynchronizationContext)continuationContext;
+
+                    if (continuationSyncCtx == Thread.CurrentThreadAssumedInitialized._synchronizationContext)
                     {
-                        throw;
+                        // Inline
+                        return false;
                     }
 
-                    continuation.GCData![(continuation.Flags & CorInfoContinuationFlags.CORINFO_CONTINUATION_RESULT_IN_GCDATA) != 0 ? 1 : 0] = ex;
-                    continue;
+                    // TODO: May need a volatile write here.
+                    TOps.SetContinuationState(task, continuation);
+
+                    try
+                    {
+                        TOps.PostToSyncContext(task, continuationSyncCtx);
+                    }
+                    catch (Exception ex)
+                    {
+                        Task.ThrowAsync(ex, targetContext: null);
+                    }
+
+                    return true;
                 }
 
-                if (newContinuation != null)
+                if ((continuation.Flags & CorInfoContinuationFlags.CORINFO_CONTINUATION_CONTINUE_ON_CAPTURED_TASK_SCHEDULER) != 0)
                 {
-                    newContinuation.Next = continuation.Next;
-                    return null;
+                    object continuationContext = continuation.GetContinuationContext();
+                    Debug.Assert(continuationContext is TaskScheduler { });
+                    TaskScheduler sched = (TaskScheduler)continuationContext;
+
+                    TOps.SetContinuationState(task, continuation);
+                    // TODO: We do not need TaskSchedulerAwaitTaskContinuation here, just need to refactor its Run method...
+                    var taskSchedCont = new TaskSchedulerAwaitTaskContinuation(sched, TOps.GetContinuationAction(task), flowExecutionContext: false);
```

Comment:
The reason we can do `flowExecutionContext: false` is because the task is always a `ThunkTask` that wraps a continuation chain and continuation resume will flow the exec context across await. Right? Maybe worth adding a comment.

Reply:
Correct. More specifically `ExecutionContext` is always saved and restored directly by the JIT generated code when necessary. So flowing the `ExecutionContext` inside the BCL would result in double save/restore of `ExecutionContext`.

Note that `ExecutionContext` handling was added separately in #117056. The handling of it is almost completely orthogonal to the continuation context handling.